server/util/log/gcp/gcp.go

```diff
@@ -0,0 +1,108 @@
+package gcp
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/logging"
+	"github.com/rs/zerolog"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	logpb "google.golang.org/genproto/googleapis/logging/v2"
+)
+
+var (
+	ProjectID = flag.String("app.log_gcp_project_id", "", "The project ID to log to in GCP (if any).")
+	LogID     = flag.String("app.log_gcp_log_id", "", "The log ID to log to in GCP (if any).")
+)
+
+func NewLogWriter() (zerolog.LevelWriter, error) {
+	if *ProjectID == "" || *LogID == "" {
+		return nil, nil
+	}
+	client, err := logging.NewClient(context.Background(), *ProjectID)
+	if err != nil {
+		return nil, err
+	}
+	return &logWriter{ctx: context.Background(), logger: client.Logger(*LogID)}, nil
+}
+
+type logWriter struct {
+	ctx    context.Context
+	logger *logging.Logger
+}
+
+func (l *logWriter) Write(p []byte) (int, error) {
+	l.logger.StandardLogger(logging.Default).Print(string(p))
+	return len(p), nil
+}
+
+func zerologLevelToGCPSeverity(l zerolog.Level) logging.Severity {
+	switch l {
+	case zerolog.DebugLevel:
+		return logging.Debug
+	case zerolog.InfoLevel:
+		return logging.Info
+	case zerolog.WarnLevel:
+		return logging.Warning
+	case zerolog.ErrorLevel:
+		return logging.Error
+	case zerolog.PanicLevel:
+		fallthrough
+	case zerolog.FatalLevel:
+		return logging.Critical
+	default:
+		return logging.Default
+	}
+}
+
+func jsonPayload(p []byte) (*structpb.Struct, error) {
+	m := map[string]any{}
+	if err := json.Unmarshal(p, &m); err != nil {
+		return nil, err
+	}
+	jsonPayload, err := structpb.NewStruct(m)
+	if err != nil {
+		return nil, err
+	}
+	return jsonPayload, nil
+}
+
+func populateEntryFromJsonPayload(entry *logging.Entry, payload *structpb.Struct) {
+	entry.Payload = payload
+
+	// Populate the Entry fields, as this won't happen automatically. The map of
+	// json to Entry fields can be found here:
+	// https://cloud.google.com/logging/docs/structured-logging
+	fields := payload.GetFields()
+	if t, err := time.Parse(zerolog.TimeFieldFormat, fields[zerolog.TimestampFieldName].GetStringValue()); err == nil {
+		entry.Timestamp = t
+	}
+	if v, ok := fields["logging.googleapis.com/sourceLocation"]; ok {
```

Comment:
Could you exclude fields from the Payload that are already present at the top level (such as this one, timestamp, severity, etc)?